stablehlo/dialect/TypeInference.cpp

```diff
@@ -97,6 +97,51 @@ bool tensorsHaveSameElType(Type type1, Type type2,
   return tensorsHaveSameElType({type1, type2}, ignoreFpPrecision);
 }
 
+unsigned potentiallyComplexBitWidth(Type type) {
+  auto complexTy = type.dyn_cast<ComplexType>();
+  return complexTy ? 2 * complexTy.getElementType().getIntOrFloatBitWidth()
+                   : type.getIntOrFloatBitWidth();
+}
+
+// Returns true if the element-type of type1 can be promoted to that of type2.
+// An element-type 'x' is promotatble to element-type 'y' is they have the same
+// base type and bitwidth(x) <= bitwidth(y). When 'x' and 'y' are quantized
+// element-types, then promotion is applied only to the 'storage_type'
+// component.
+bool isPromotableElementType(Type type1, Type type2,
+                             bool ignoreFpPrecision = false) {
+  auto tensorTy1 = type1.dyn_cast<ShapedType>();
+  auto tensorTy2 = type2.dyn_cast<ShapedType>();
+
+  if (!tensorTy1 || !tensorTy2) return false;
+
+  Type tensorEl1 = tensorTy1.getElementType();
+  Type tensorEl2 = tensorTy2.getElementType();
+
+  if (ignoreFpPrecision && tensorEl1.isa<FloatType>() &&
+      tensorEl2.isa<FloatType>())
+    return true;
+
+  bool isSameType =
+      (tensorEl1.isa<IntegerType>() and tensorEl2.isa<IntegerType>()) ||
+      (tensorEl1.isa<FloatType>() and tensorEl2.isa<FloatType>()) ||
+      (tensorEl1.isa<ComplexType>() and tensorEl2.isa<ComplexType>()) ||
+      (tensorEl1.isa<quant::QuantizedType>() and
+       tensorEl2.isa<quant::QuantizedType>());
```

Comment:
A short helper method may help for all these:

```
template <typename T>
bool matchesType(Type a, Type b) {
   return a.isa<T>() && b.isa<T>();
}
```

Reply:
done!